wpe/src/main/cpp/browser/messagepump.cpp

```diff
@@ -0,0 +1,226 @@
+    while (it!=m_looperGlibPollFds.end()) {
+        ALooper_removeFd(m_looper, it->fd);
+        it = m_looperGlibPollFds.erase(it);
+    }
+
+    g_free(m_pollFds);
+    m_pollFdsSize = 0;
+
+    /* Release GMainContext loop */
+    g_main_context_unref(m_context);
+
+    ALooper_release(m_looper);
+    m_looper = nullptr;
+
+    close(m_dispatchFd);
+}
+
+void MessagePump::invoke(void (* callback)(void*), void* callbackData, void (* destroy)(void*))
+{
+    struct GenericCallback
+    {
+        void (* callback)(void*);
+        void* callbackData;
+        void (* destroy)(void*);
+    };
+
+    auto* data = new GenericCallback { callback, callbackData, destroy };
+    g_main_context_invoke_full(m_context, G_PRIORITY_DEFAULT, +[](gpointer data) -> gboolean {
+        auto* genericData = reinterpret_cast<GenericCallback*>(data);
+        genericData->callback(genericData->callbackData);
+        return G_SOURCE_REMOVE;
+    }, data, +[](gpointer data) -> void {
+        auto* genericData = reinterpret_cast<GenericCallback*>(data);
+        if (genericData->destroy)
+            genericData->destroy(genericData->callbackData);
+        delete genericData;
+    });
+}
+
+void MessagePump::prepare() {
+    g_main_context_prepare (m_context, &m_maxPriority);
+
+    if (m_pollFds == nullptr) {
+        m_pollFdsSize = 1; // There will be at least one fd in GMainContext
+        m_pollFds = g_new (GPollFD, m_pollFdsSize);
+    }
+
+    gint nfds;
+    gint timeout = 0;
+    while ((nfds = g_main_context_query (m_context, m_maxPriority, &timeout,
+                                         m_pollFds,
+                                         m_pollFdsSize)) > m_pollFdsSize) {
+        g_free(m_pollFds);
+        m_pollFdsSize = nfds;
+        m_pollFds = g_new (GPollFD, nfds);
+    }
+
+    char *glibPollFdAddedToLooperFlags = (char *)g_new0(char, nfds);
+
+    // Update reference count of each looper glib poll fd
+    for (std::vector<LooperGLibPollFd>::iterator it=m_looperGlibPollFds.begin(); it!=m_looperGlibPollFds.end(); ++it) {
+        it->ref = 0;
+
+        for (int i=0; i < m_pollFdsSize; i++) {
+            GPollFD* pfd = m_pollFds+i;
+
+            if (it->fd == pfd->fd) {
+                *(glibPollFdAddedToLooperFlags+i) = 1;
+                it->pfd = pfd;
+                it->ref = 1;
+                pfd->revents = 0;
+                break;
+            }
+        }
+    }
+
+    // Add new fds to ALooper if they weren't added before
+    for (int i=0; i < m_pollFdsSize; i++) {
+        GPollFD *pfd = m_pollFds + i;
+        gint exists = (gint) *(glibPollFdAddedToLooperFlags + i);
+
+        if (exists)
+            continue;
+
+        pfd->revents = 0;
+
+        // ALooper uses epoll for polling and in order to dispatching work correctly it requires ALOOPER_EVENT_OUTPUT
+        // but G_IO_OUT is not flagged by default glib wakup fd nor by wpe android backend ipc sockets.
```

Comment:
Typo: `wakup` → `wakeup`

```suggestion
        // but G_IO_OUT is not flagged by default glib wakeup fd nor by wpe android backend ipc sockets.
```

Reply:
fixed